### PR TITLE
MQE: add docs about native histogram memory pooling and safety

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR expands the readme for `pkg/streamingpromql` to explain how `histogram.FloatHistogram` instances are reused and what should be done to ensure this happens correctly and safely.

I've also taken this opportunity to update references to "the streaming PromQL engine" to "MQE" in the readme.

I have not added a changelog entry given this is not a user-facing change.

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [n/a] Tests updated.
- [x] Documentation added.
- [n/a] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [n/a] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/streamingpromql/README.md [hunk 1]
```diff
@@ -98,4 +98,25 @@ Elaborating on the example from before, the overall query would proceed like thi
...
+
+MQE makes extensive use of memory pooling to reduce GC pressure, including for slices that hold `*histogram.FloatHistogram` pointers.
+
+Slices of `promql.HPoint` returned by `types.HPointSlicePool` are not cleared when they are returned. This allows the `FloatHistogram`
+instances to be reused for other series or time steps. For example, when filling a `HPointRingBuffer`, range vector selectors will
+reuse `FloatHistogram` instances already present in the `HPoint` slice that backs the ring buffer, rather than unconditionally creating
+a new `FloatHistogram` instance.
+
+The implication of this is that anywhere that returns a `promql.HPoint` slice `s` to `types.HPointSlicePool` must remove references in
+`s` to any `FloatHistogram`s retained after `s` is returned. The simplest way to do this is to set the `H` field on `promql.HPoint` to `nil`.
```
Comment (jhesketh) [inline pkg/streamingpromql/README.md:115]: An explanation of why you may return `s` if you are retaining histograms could be helpful. eg when wouldn't I keep all the points and the slice?

File: pkg/streamingpromql/README.md [hunk 2]
```diff
@@ -98,4 +98,25 @@ Elaborating on the example from before, the overall query would proceed like thi
...
+If all points in `s` are being retained, then calling `clear(s)` is also sufficient.
+
+If this is not done, query results may become corrupted due to multiple queries simultaneously modifying the same `FloatHistogram` instance.
+This can also manifest as panics while interacting with `FloatHistogram`s.
+
+The same problem does not apply to `*histogram.FloatHistogram` slices returned by `types.HistogramSlicePool`. Slices from this pool are used only by
+parts of MQE that do not benefit from reusing `FloatHistogram` instances, and so `types.HistogramSlicePool` clears all slices returned.
```
Comment (jhesketh) [inline pkg/streamingpromql/README.md:122]:
```suggestion
parts of MQE that do not benefit from reusing `FloatHistogram` instances, and so `types.HistogramSlicePool` clears all slices returned for you.
```

**Review by jhesketh — APPROVED** [review]
lgtm, just some nits

File: pkg/streamingpromql/README.md (hunk 1 above)
Comment (charleskorn) [reply pkg/streamingpromql/README.md]: Good idea, added in c043e5b.

**Pull request merged**